isamples_metadata/OpenContextTransformer.py

```diff
@@ -195,30 +201,79 @@ def sample_sampling_purpose(self) -> str:
+            return self._material_prediction_results
+
+    def has_material_categories(self) -> typing.List[str]:
+        item_category = self.source_record.get("item category", "")
+        to_classify_items = ["Object", "Pottery", "Sample", "Sculpture"]
+        if item_category in to_classify_items:
+            prediction_results = self._compute_material_prediction_results()
+            if prediction_results is not None:
+                return [prediction.value for prediction in prediction_results]
+            else:
+                return []
         return MaterialCategoryMetaMapper.categories(item_category)
 
+    def has_material_category_confidences(self, material_categories: list[str]) -> typing.Optional[typing.List[float]]:
+        prediction_results = self._compute_material_prediction_results()
+        if prediction_results is None:
+            return None
+        else:
+            return [prediction.confidence for prediction in prediction_results]
+
+    def _compute_specimen_prediction_results(self) -> typing.Optional[typing.List[PredictionResult]]:
+        item_category = self.source_record.get("item category", "")
+        to_classify_items = ["Animal Bone"]
+        if item_category not in to_classify_items:
+            # Have specified mapping, won't predict
+            return None
+        elif self._specimen_prediction_results is not None:
+            # Have already computed, don't predict again
+            return self._specimen_prediction_results
+        else:
+            # Need to predict
+            # get the model
+            ocs_model = MetadataModelLoader.get_oc_sample_model()
+            # load the model predictor
+            ocsp = OpenContextSamplePredictor(ocs_model)
+            self._specimen_prediction_results = ocsp.predict_sample_type(self.source_record)
+            return self._specimen_prediction_results
+
     def has_specimen_categories(self) -> typing.List[str]:
         item_category = self.source_record.get("item category", "")
         to_classify_items = ["Animal Bone"]
         if item_category in to_classify_items:
-            # we assume item category value exists for every record
-            # if item category value is ambiguous (has multiple possible material labels)
-            # invoke the machine learning model
-            ocm_model = MetadataModelLoader.get_oc_sample_model()
-            ocsp = OpenContextSamplePredictor(ocm_model)
-            return [prediction.value for prediction in ocsp.predict_sample_type(self.source_record)]
+            prediction_results = self._compute_specimen_prediction_results()
+            if prediction_results is not None:
+                return [prediction.value for prediction in prediction_results]
+            else:
+                return []
         return SpecimenCategoryMetaMapper.categories(item_category)
 
+    def has_specimen_category_confidences(self, specimen_categories: list[str]) -> typing.Optional[typing.List[float]]:
```

Comment:
add confidence checking for specimen and material. This is based on Danny's work in `SESARTransformer`.